fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java

```diff
@@ -0,0 +1,246 @@
+		if (EventFactory.isProfilingEnabled()) {
+			final Profiler profiler = world.getProfiler();
+			profiler.push("fabricServerBlockEntityUnload");
+
+			for (BlockEntityUnloadCallback<ServerWorld> callback : callbacks) {
+				profiler.push(EventFactory.getHandlerName(callback));
+				callback.onUnloadBlockEntity(blockEntity, world);
+				profiler.pop();
+			}
+
+			profiler.pop();
+		} else {
+			for (BlockEntityUnloadCallback<ServerWorld> callback : callbacks) {
+				callback.onUnloadBlockEntity(blockEntity, world);
+			}
+		}
+	});
+
+	/**
+	 * Called when an Entity is loaded into a ServerWorld.
+	 *
+	 * <p>Note there is no corresponding unload event because entity unloads cannot be reliably tracked.
+	 */
+	public static final Event<EntityLoadCallback<ServerWorld>> ENTITY_LOAD = EventFactory.createArrayBacked(EntityLoadCallback.class, callbacks -> (entity, world) -> {
+		if (EventFactory.isProfilingEnabled()) {
+			final Profiler profiler = world.getProfiler();
+			profiler.push("fabricServerEntityLoad");
+
+			for (EntityLoadCallback<ServerWorld> callback : callbacks) {
+				profiler.push(EventFactory.getHandlerName(callback));
+				callback.onEntityLoad(entity, world);
+				profiler.pop();
+			}
+
+			profiler.pop();
+		} else {
+			for (EntityLoadCallback<ServerWorld> callback : callbacks) {
+				callback.onEntityLoad(entity, world);
+			}
+		}
+	});
+
+	/**
+	 * Called when the server has started. At this stage, all worlds are live.
+	 */
+	public static final Event<GameLifecycleCallback<MinecraftServer>> SERVER_STARTED = EventFactory.createArrayBacked(GameLifecycleCallback.class, (callbacks) -> (server) -> {
+		for (GameLifecycleCallback<MinecraftServer> callback : callbacks) {
+			callback.onChangeLifecycle(server);
+		}
+	});
+
+	/**
+	 * Called when the server has started stopping. All worlds are still present.
+	 */
+	public static final Event<GameLifecycleCallback<MinecraftServer>> SERVER_STOPPING = EventFactory.createArrayBacked(GameLifecycleCallback.class, (callbacks) -> (server) -> {
+		for (GameLifecycleCallback<MinecraftServer> callback : callbacks) {
+			callback.onChangeLifecycle(server);
+		}
+	});
+
+	/**
+	 * Called when the server has stopped. All worlds have been closed and all (block)entities and players have been unloaded.
+	 */
+	public static final Event<GameLifecycleCallback<MinecraftServer>> SERVER_STOPPED = EventFactory.createArrayBacked(GameLifecycleCallback.class, callbacks -> server -> {
+		for (GameLifecycleCallback<MinecraftServer> callback : callbacks) {
+			callback.onChangeLifecycle(server);
+		}
+	});
+
+	/**
+	 * Gets the currently running server.
+	 *
+	 * <p>The server instance returned SHOULD NOT be cached! Call the method every time you need the server.
+	 *
+	 * @return the currently running server
+	 * @throws IllegalStateException if the server is not available
+	 */
+	public static MinecraftServer getCurrentServer() {
```

Comment:
While this is useful and sometimes not truly avoidable, I am not sure if we want to reinforce the notion that there's only one server. There should at least be a strong recommendation to always use world.getServer and other context specific approaches instead.

Reply:
Yea that is precisely why I did stage this addition to one commit incase it was unpopular. The documentation would be nice to have about reccomending getting the server from the ServerWorld.

Internally this will optionally get the depending on a client or server. Which you can look at in impl packages.

Reply:
Hmm, I think this might `getPrimaryServer` might be a better name since we don't assume one server exists. I've added some docs explaining how it's use is discouraged.